### ISSUE TITLE
Validators have more than double the expected number of append vecs

### ISSUE THREAD

**Issue by brooksprumo** [solana-labs/solana#26259]
#### Problem

Some validators running master/v1.11 are hitting mmap/open file limits. Upon investigation on mce2, there are 900k+ append vecs in `${ledger}/accounts/accounts/`, which is the directory that snapshots archives are inflated into. 

Why are there so many append vecs in the snapshot?
Why are they not getting cleaned up/shrunk?
How were they created in the first place?

#### Proposed Solution

Root-cause and fix.

**Comment by brooksprumo** [comment]
Also observed by @HaoranYi on dev2. Also running master/v1.11.

```
sol@dev-server-ny5:~$ ls  /mnt/nvme1n1/ledger/accounts | wc -l
25706
sol@dev-server-ny5:~$ ls  /mnt/nvme1n1/ledger/accounts/accounts/ | wc -l
902325
```

**Comment by brooksprumo** [comment]
On `mce2`, the snapshot sizes look quite reasonable: 

```
-rw-r--r--  1 sol sol  32G Jun 27 08:32 snapshot-139211562-J1m9YHTPmQUvt1VjbyNp4hFFEgJPEurLto7AGFHo5pxg.tar.zst
-rw-r--r--  1 sol sol  32G Jun 27 13:57 snapshot-139240745-D17vR2iksG5RoLMfTX7i5NwSsr4VpbybuX1eqzesQfu2.tar.zst


32 GB for the full snapshots, and 2 GB for the incremental snapshots.

Opening up 
-rw-r--r--  1 sol sol  32G Jun 27 13:57 snapshot-139240745-D17vR2iksG5RoLMfTX7i5NwSsr4VpbybuX1eqzesQfu2.tar.zst

**Comment by jeffwashington** [comment]
hmm. if we pass `--accounts-db-skip-rewrites` to a validator, we will accumulate old append vecs. Rent rewrites normally allow us to clean up old append vecs since all accounts in them will be rewritten within the most recent 432k slots.
However, it has been weeks(?) since I ran a validator with only the skip rewrites option and not ALSO running with `--accounts-db-ancient-append-vecs`. Maybe there is a bare metal validator still running with only skip rewrites and a snapshot came from one of those.

**Comment by jeffwashington** [comment]
or maybe another validator is skipping shrink (or otherwise failing to shrink) like the bug I'm chasing and we're accumulating append vecs and then it makes a snapshot. That would be consistent behavior with the bug I'm chasing.

**Comment by brooksprumo** [comment]
`solana-validator` cli args from `mce2`:

```
solana-validator
--dynamic-port-range 8002-8020
--gossip-port 8001
--identity /home/sol/identity/validator-identity-am6-2.json
--ledger /home/sol/ledger
--limit-ledger-size
--log /home/sol/logs/solana-validator.log
--rpc-port 8899
--expected-genesis-hash 5eykt4UsFv8P8NJdTREpY1vzqKqZKvdpKuc147dw2N9d
--wal-recovery-mode skip_any_corrupted_record
--public-rpc-address 147.75.80.133:80
--no-voting
--trusted-validator ba7p14iARCNfq8wrcAakfNXrqY4CtJ2C4eE5ic79mpF 
--trusted-validator 7Np41oeYqPefeNQEHSv1UDhYrehxin3NStELsSKCT4K2 
--trusted-validator GdnSyH3YtwcxFvQrVVJMm1JhTS4QVX7MFsX56uJLUfiZ 
--trusted-validator DE1bawNcRJB9rVm3buyMVfr8mBEoyyu73NBovf2oXJsJ 
--trusted-validator CakcnaRDHka2gXyfbEd2d3xsvkJkqsLw2akB3zsN1D2S 
--halt-on-trusted-validators-accounts-hash-mismatch 
--no-untrusted-rpc 
--entrypoint mainnet-beta.solana.com:8001 
--entrypoint entrypoint2.mainnet-beta.solana.com:8001 
--entrypoint entrypoint3.mainnet-beta.solana.com:8001 
--entrypoint entrypoint4.mainnet-beta.solana.com:8001 
--entrypoint entrypoint5.mainnet-beta.solana.com:8001 
--no-genesis-fetch 
--no-snapshot-fetch 
--accounts /mnt/accounts 
--entrypoint entrypoint2.mainnet-beta.solana.com:8001 
--entrypoint entrypoint3.mainnet-beta.solana.com:8001 
--entrypoint entrypoint4.mainnet-beta.solana.com:8001 
--entrypoint entrypoint5.mainnet-beta.solana.com:8001
```

**Comment by t-nelson** [comment]
fwiw `mce2` is _generating_ these snapshots.  when i first encountered it last week, `mce2` was ~200k slots behind and had be bootlooping on snapshot extraction (OOM/maps exhausted).  it's unclear whether the snapshot that initiated this behavior was generated locally or downloaded.  command line args would suggest local, but there's no audit log on how `mce2` had been administered

**Comment by jeffwashington** [comment]
@t-nelson how did you detect this:
(OOM/maps exhausted)
Did it show up as ram used or as an oom panic somehow when we tried to mmap?

**Comment by HaoranYi** [comment]
> if it is master, then downloaded snapshots would be in /remote locally generated snapshots are next to /remote @HaoranYi

yes. locally generate snapshots are in `$snapshotdir`; remotely downloaded snapshot are in `$snapshotdir/remote`.

**Comment by brooksprumo** [comment]
> fwiw `mce2` is _generating_ these snapshots. 

Yep. Generating these _new_ snapshots. It's possible this node downloaded a huge-append-vec snapshot to begin with, but we cannot tell now. In either case, the node doesn't seem to be shrinking/cleaning up the append vecs, so something during steady state is up.

**Comment by steviez** [comment]
Another datapoint from `dev11`; I have the following snapshot which has a reasonable looking compressed file size (in line with what Brooks posted above):
```
-rw-r--r-- 1 sol users  31G Jun 18 17:26 snapshot-138039141-BFgoh4yjfSFFDJB6vcH8Rrj9Bi9GNLVBbRpUXz4m34Qa.tar.zst
```
After it has been expanded by `solana-ledger-tool verify`, I see it has the following number of files (higher than what Brooks reported as "normal" above, but not as bad a the 1.4M number he saw on one snapshot):
```
$ ls ~/ledger/accounts.ledger-tool/accounts/ | wc -l
372515
```
However, the reported disk footprint looks way too big:
```
$ du -sh ~/ledger/accounts.ledger-tool/
203G	/home/sol/ledger/accounts.ledger-tool/
```
I know this expanded size has been growing as we pick up more accounts, but I believe it should be on the order of 90G - 100G. This was with several day old master, and all default args with ledger-tool verify (ie no extra flags).

**Comment by jeffwashington** [comment]
I also saw a 200G accounts folder. And that folder was a ram disk, so the 200G of files were occupying 200G of ram.

**Comment by HaoranYi** [comment]
The total size of the accounts dir seems to be reasonable in my run.
```
sol@dev-server-ny5:~$ du -hs /mnt/nvme1n1/ledger/accounts/accounts/
55G     /mnt/nvme1n1/ledger/accounts/accounts/
sol@dev-server-ny5:~$ du -hs /mnt/nvme1n1/ledger/accounts/
102G    /mnt/nvme1n1/ledger/accounts/
```

**Comment by brooksprumo** [comment]
> The total size of the accounts dir seems to be reasonable in my run.
> 
> ```
> sol@dev-server-ny5:~$ du -hs /mnt/nvme1n1/ledger/accounts/accounts/
> 55G     /mnt/nvme1n1/ledger/accounts/accounts/
> sol@dev-server-ny5:~$ du -hs /mnt/nvme1n1/ledger/accounts/
> 102G    /mnt/nvme1n1/ledger/accounts/
> ```

What are the number of files in the directories?

**Comment by jeffwashington** [comment]
I think ~100G total is 'correct' for mnb atm.

**Comment by jeffwashington** [comment]
> I think ~100G total is 'correct' for mnb atm.

I checked 3 bare metal validators running against master right now and making roots.
Sizes of ledger/accounts* was 92G, 94G, 97G.

**Comment by jeffwashington** [comment]
on dev11/sg1-2, startup on a hacked up 1.10 ledger tool reports 31.7M duplicate pubkeys in the append vecs from the snapshot.

**Comment by jeffwashington** [comment]
```
here are the slots present in append vecs for the pubkey with most duplicate data len:
7CLo1BY41BHAVnEs57kzYMnWXyBJrVEBPpZyQyPo2p1G, slots: 138038606, 138038613, 138038614, 138038620, 138038623,
138038627, 138038634, 138038638, 138038640, 138038648, 138038649, 138038652, 138038660, 138038661, 138038664,
138038668, 138038672, 138038678, 138038681, 138038685, 138038689, 138038694, 138038697, 138038704, 138038708,
138038712, 138038716, 138038721, 138038724, 138038727, 138038732, 138038736, 138038738, 138038741, 138038745,
138038760, 138038762, 138038765, 138038769, 138038774, 138038776, 138038780, 138038785, 138038788, 138038797,
138038800, 138038804, 138038809, 138038816, 138038820, 138038828, 138038831, 138038835, 138038838, 138038840,
138038843, 138038849, 138038850, 138038854, 138038856, 138038860, 138038866, 138038869, 138038872, 138038876,
138038880, 138038884, 138038888, 138038896, 138038898, 138038901, 138038904, 138038908, 138038912, 138038917,
138038922, 138038927, 138038930, 138038935, 138038941, 138038947, 138038948, 138038953, 138038956, 138038958,
138038964, 138038965, 138038969, 138038971, 138038977, 138038984, 138038988, 138038993, 138038997, 138039000,
138039004, 138039009, 138039013, 138039019, 138039024, 138039028, 138039032, 138039037, 138039042, 138039047,
138039052, 138039056, 138039062, 138039069, 138039070, 138039073, 138039079, 138039080, 138039083, 138039088,
138039096, 138039098, 138039102, 138039106, 138039110, 138039112, 138039120, 138039124, 138039127, 138039132,
138039137, 138039140, 138050668, 138050673, 138050678, 138050682, 138050688, 138050692, 138050695, 138050700,
138050709, 138050712, 138050716, 138050717, 138050726, 138050732, 138050735, 138050736, 138050751, 138050752,
138050756, 138050760, 138050768, 138050772, 138050777, 138050780, 138050785, 138050793, 138050805, 138050806,
138050811, 138050815, 138050819, 138050820, 138050824, 138050827, 138050839, 138050841, 138050844, 138050853,
138050854, 138050858, 138050860, 138050868, 138050872, 138050876, 138050881, 138050887, 138050888, 138050892,
138050898, 138050901, 138050905, 138050914, 138050915, 138050919, 138050924, 138050926, 138050930, 138050936,
138050940, 138050945, 138050951, 138050964, 138050965, 138050968, 138050972, 138050974, 138050981, 138050986, 
138050989, 138050995, 138051003, 138051009, 138051014, 138051019, 138051023, 138051028, 138051033, 138051037, 
138051045, 138051046, 138051052, 138051056, 138051063, 138051064, 138051074, 138051075, 138051079, 138051081, 
138051084, 138051090, 138051095, 138051101, 138051107, 138051108, 138051112, 138051117, 138051121, 138051128, 
138051132, 138051144, 138051148, 138051153, 138051156, 138051163, 138051166, 138051171, 138051179, 138051181, 
138051184, 138051190, 138051192, 138051198, 138051202, 138051206, 138051209, 138051213, 138051218, 138051224, 
138051228, 138051232, 138051236, 138051240, 138051247, 138051248, 138051252, 138051259, 138051263, 138051267, 
138051272, 138051277, 138051283, 138051287, 138051292, 138051295, 138051301, 138051302, 138051304, 138051306, 
138051314, 138051317, 138051324, 138051329, 138051335, 138051339, 138051343, 138051347, 138051356, 138051360, 
138051362, 138051369, 138051370, 138051375, 138051380, 138051383, 138051387, 138051392, 138051397, 138051404, 
138051411, 138051416, 138051418, 138051423, 138051429, 138051433, 138051439, 138051444, 138051448, 138051457, 
138051460, 138051465, 138051469, 138051474, 138051480, 138051488, 138051492, 138051500, 138051502, 138051507, 
138051512, 138051518, 138051524, 138051528, 138051532, 138051536, 138051544, 138051546, 138051549, 138051554, 
138051566, 138051567, 138051573, 138051578, 138051582, 138051588, 138051592, 138051597, 138051603, 138051608, 
138051611, 138051616, 138051621, 138051626, 138051629, 138051633, 138051636, 138051642, 138051647, 138051648, 
138051652, 138051657, 138051668, 138051672, 138051678, 138051680, 138051688, 138051696, 138051701, 138051707,
138051712, 138051717, 138051720, 138051724, 138051728, 138051739, 138051741, 138051746, 138051754, 138051755,
138051756, 138051760, 138051762, 138051774, 138051776, 138051785, 138051789, 138051791, 138051794, 138051803,
138051808, 138051812, 138051819, 138051826, 138051833, 138051836, 138051844, 138051845, 138051848, 138051853,
138051858, 138051862, 138051868, 138051872, 138051875, 138051878, 138051884, 138051888, 138051901, 138051908,
138051916, 138051922, 138051925, 138051927, 138051941, 138051943, 138051944, 138051952, 138051955, 138051960, 
138051963, 138051973, 138051980, 138051983, 138051992, 138052000, 138052004, 138052008, 138052018, 138052026,
138052032, 138052034, 138052039, 138052040, 138052048, 138052051, 138052056, 138052061, 138052063, 138052069, 
138052074, 138052078, 138052080, 138052082, 138052085, 138052089, 138052095, 138052098, 138052102, 138052107, 
138052111, 138052122, 138052126, 138052131, 138052134, 138052137, 138052148, 138052152, 138052156, 138052160, 
138052168, 138052174, 138052179, 138052182, 138052192, 138052197, 138052201, 138052205, 138052211, 138052214, 
138052217, 138052220, 138052227, 138052228, 138052232, 138052237, 138052249, 138052254, 138052256, 138052264, 
138052267, 138052280, 138052288, 138052291, 138052293, 138052298, 138052300, 138052306, 138052309, 138052316, 
138052324, 138052328, 138052336, 138052340, 138052352, 138052353, 138052356, 138052363, 138052364, 138052367, 
138052374, 138052376, 138052380, 138052385, 138052389, 138052390, 138052400, 138052403, 138052410, 138052421, 
138052422, 138052424, 138052428, 138052432, 138052435, 138052439, 138052443, 138052448, 138052452, 138052456,
138052460, 138052464, 138052472, 138052477, 138052480, 138052482, 138052487, 138052493, 138052495, 138052500,
138052504, 138052506, 138052512, 138052516, 138052517, 138052520, 138052527, 138052528, 138052531, 138052536, 
138052540, 138052544, 138052548, 138052560, 138052562, 138052568, 138052579, 138052584, 138052586, 138052591, 
138052595, 138052599, 138052605, 138052609, 138052611, 138052624, 138052628, 138052633, 138052636, 138052643, 
138052647, 138052652, 138052655, 138052663, 138052670, 138052671, 138052674, 138052679, 138052688, 138052696,
138052700, 138052705, 138052707, 138052710, 138052719
each entry has data len: 6,720,200
duplicate pubkeys: 31,681,391, data len from duplicates: 123,144,404,264, original data len: 181,663,654,649
20 highest duplicate pubkey data size sum
(sum of data in duplicate append vec entries, pubkey)
(3689389800, 7CLo1BY41BHAVnEs57kzYMnWXyBJrVEBPpZyQyPo2p1G)
(3649068600, EH32v4UHcwH6S7gLTRvEBEyCTJrVbhRiJE7QEGoqd4NU)
(3649068600, CGmWwBNsTRDENT5gmVZzRu38GnNnMm1K5C3sFiUUyYQX)
(3642348400, 5WyTBrEgvkAXjTdYTLY9PVrztjmz4edP5W9wks9KPFg5)
(3615467600, 2v4qnkskmnXjeYRQkdFTPLPsU42vhG1tAKbRvw7Vqjij)
(3588586800, CcPVS9bqyXbD9cLnTbhhHazLsrua8QMFUHTutPtjyDzq)
(3561706000, 76npM99oWkDXdepEJLXc3chmya2n1tEZzqfU2n67nywS)
(2163237044, 6NQqaa48SnBBJZt9HyVPngcZFW81JfDv9EjRX2M4WkbP)
(2119196348, 3z4QQPFdgNSxazqEAzmZD5C5tJWepczimVqWak2ZPY8v)
(2112904820, GR363LDmwe25NZQMGtD2uvsiX66FzYByeQLcNFr596FK)
(1950373680, Eac7hqpaZxiBtG4MdyKpsgzcoVN6eMe9tAbsdZRYH4us)
(1742753256, 2XHxua6ZaPKpCGUNvSvTwc9teJBmexp8iMWCLu4mtzGb)
(1604339640, GRtx6o9nipeoGqhKFUygUotf9FZncT4vqSqtUu3ZWYmW)
(1521501188, 6o44a9xdzKKDNY7Ff2Qb129mktWbsCT4vKJcg2uk41uy)
(1516258248, CTZuXPjhrLb4PSNSqdsc7xUn8eiRAByfQXoi4HXkPVUe)
(1416642388, RC7ygxnwYpqNETVnTVzhqnzWhGata8bccCZB5dcYoH5)
(1379941808, FERWWtsZoSLcHVpfDnEBnUqHv4757kTUUZhLKBCbNfpS)
(1295006180, 75ZYEwx8X9q625m4qFWK6bsVrRbAxHFbK18hyFos7YiJ)
(1252014072, 2i34Kriz23ZaQaJK6FVhzkfLhQj8DSqdQTmMwz4FF9Cf)
(1172321384, DK7Jw2fZCbTVDF191eXywDFymmbXqGUp4VuERiX6RAR4)
accounts data len: 58,519,250,385
```

**Comment by brooksprumo** [comment]
@jeffwashington For the accounts with lots of duplicates, do you know if they are all programs (`is_executable == true`)?

**Comment by jeffwashington** [comment]
Here is part of the issue:
#25936
I put this in because it was not possible to start from a snapshot with many accounts because of oom during clean.
This is related to the fact we don't clean or shrink when we load a snapshot that was generated on this machine.
The assumption is that the snapshot was being cleaned and shrunk during steady state runtime.
So, this is why we don't clean the snapshot up as aggressively as we used to if the snapshot was generated on this machine.

This assumption (that we were cleaning and shrinking during steady state) is invalid if we're not cleaning and shrinking during steady state...
So, that still appears to be the root of this issue - that we're either no cleaning or not shrinking correctly during steady state.

**Comment by jeffwashington** [comment]
Related to this is @HaoranYi's change (that is in 1.10, too) that if the snapshot is not in /remote, then we don't clean or shrink at startup.

**Comment by steviez** [comment]
The top 3 from the list jwash provided all have "Chainlink Store Program" as their "Assigned Program Id"; for example,
https://explorer.solana.com/address/7CLo1BY41BHAVnEs57kzYMnWXyBJrVEBPpZyQyPo2p1G

**Comment by HaoranYi** [comment]
Looking at `7CLo1BY41BHAVnEs57kzYMnWXyBJrVEBPpZyQyPo2p1G` on the explorer, there are lots of transactions on this account, A.K.A 5 transactions within one slot.  And many of transactions failed.

**Comment by diman-io** [comment]
JFI
(I didn't see another similar messages in Discord. So I don't know, may be it's my config bug)

First time message https://discord.com/channels/428295358100013066/689412830075551748/955742070780330014

After that I changed machine/os
Now it is Ubuntu 22.04 LTS (GNU/Linux 5.15.0-27-generic x86_64)

```
# ls /mnt/ramdisk/solana/accounts/ | wc -l
898981
# ls /mnt/ramdisk/solana/accounts/accounts | wc -l
11437
# cp -f /proc/`pidof solana-validator`/maps ~/maps.txt; sed -e 's/^.* \(\/[^ ]\+\)\/[^\/]\+$/\1/;t;d' maps.txt | awk '{count[$0]++} END{for (i in count) print count[i], i}'
898857 /mnt/ramdisk/solana/accounts
3 /root/.local/share/solana/install/releases/1.10.26/solana-release/bin
4 /root/.local/share/solana/install/releases/1.10.26/solana-release/bin/perf-libs
46 /usr/lib/x86_64-linux-gnu
11434 /mnt/ramdisk/solana/accounts/accounts
# sysctl vm.max_map_count
vm.max_map_count = 1000000
```

solana.service:
```
...
LimitNOFILE=1024000
ExecStart=/opt/solana/validator/validator.sh
...
```
---
<img width="631" alt="image" src="https://user-images.githubusercontent.com/71597545/176259889-52eefd4d-2851-4a3e-b646-83fb887e81a2.png">
Drops on the chart are restarts.

```
openfiles=$(cat /proc/sys/fs/file-nr | awk '{ print $1 }')
```

**Comment by brooksprumo** [comment]
@diman-io Thanks for this info. It looks like from your graph that you start from a snapshot with a "normal" number of append vecs, and then at some point the number starts growing significantly faster. Do you happen to have validator logs from around those inflection points?

**Comment by jeffwashington** [comment]
on sg1-2, the snapshot creates 203G worth of append vecs. When I run ledger-tool and it shrinks at startup, the ending accounts folder correctly contains 94G of append vecs. This demonstrates that shrink is working as expected in some conditions. Shrink at startup is thorough.

**Comment by steviez** [comment]
I started writing up another issue, but then realized that the underlying issue is likely related or the same as the one here. So, here is another datapoint that may or may not be useful.

In the process of debugging a different issue, I've been creating snapshots with different versions. The full snapshot that I started with is 1008M; this snapshot came from someone operating a testnet node with `v1.10.27`. This file size is consistent with sizes I see from our testnet warehouse google bucket.
```
-rw-r--r-- 1 ... 1008M Jul  1 17:15 snapshot-138726774-58YDKPX6MBzQBG79vpfbNd5xqDD77r5nT7PgiGdwvvpo.tar.zst
```
I created a full snapshot with `v1.10.27` that advanced forward about ~46k slots:
```
-rw-rw-r-- 1 ... 1.3G Jun 30 19:46 snapshot-138773288-agCF5RhnoN76u7pHYS761NooAAifWuuy7L4SG7eeZ8a.tar.zst
```
I also created a full snapshot with the tip of master 88b5b7e30a0e94d1108ff6bf0a89756fc41cf2e3 to the same slot:
```
-rw-rw-r-- 1 ... 9.0G Jul  2 01:31 snapshot-138773288-agCF5RhnoN76u7pHYS761NooAAifWuuy7L4SG7eeZ8a.tar.zst
```
Both snapshot were created with the default ledger-tool args:
```
solana-ledger-tool create-snapshot 138773288 --ledger /my/ledger/path
```
